setpars usage

bwgref opened the pull request:
Addresses #2 #3 and #4.

You can take a look at decide if you want to adopt these changes.

Used setPars PyXspec formalism here rather than accessing the parameters directly in the models. For me this sped things up considerably (many minutes --> seconds) for the creation of the bgdparams.xcm file, especially when using spectrum from multiple background regions.

Also forced all of the line strengths to be relative and tied together rather than independent.

This only really matters for the intbgd since it has a bajillion lines in the model, so I only did it there. Might be further speed improvements if you want to propagate this throughout.

bwgref committed:
Updated intn norms

Just reduced presets by 1e-4

bwgref commented:
Okay, don't merge this. Bug found.

achronal commented on nuskybgd/model.py:
```diff
@@ -523,71 +523,138 @@ def applymodel_intbgd(presets, refspec, bgddetimsum, model_num, src_number=None,
-                m_line.LineE.values = mod_intbgd[attr][fpm]['linee']
-                m_line.Width.values = mod_intbgd[attr][fpm]['width']
-                m_line.LineE.frozen = True
-                m_line.Width.frozen = True
-                m_line.norm.values = np.sum(
+                line_norm = np.sum(
                     bgddetimsum[spec_arrinx] * np.array(
                         mod_intbgd[attr][fpm]['ifactors'])
                 ) / np.sum(bgddetimsum[spec_arrinx])
+                
+                if not ref_set:
+                    ref_set = True
+                    ref_norm = line_norm
+                    ref_par = int(parnum+2) + m_ref_npar_offset
+                    this_par = {parnum:f"{mod_intbgd[attr][fpm]['linee']}, -0.1",
+                            parnum+1:f"{mod_intbgd[attr][fpm]['width']}, -0.1",
+                            parnum+2:f"{line_norm}, 0.1"}
+                else:
+                    rel_norm = line_norm / ref_norm                
+                    this_par = {parnum:f"{mod_intbgd[attr][fpm]['linee']}, -0.1",
+                                parnum+1:f"{mod_intbgd[attr][fpm]['width']}, -0.1",
+                                parnum+2:f"={model_name}:{ref_par}*{rel_norm}"}
+                par_dict.update(this_par)
+                parnum += 3
+
+                continue
+                # Below is the old way to do this
+#                 m_line = getattr(m, attr)
+#                 m_line.LineE.values = mod_intbgd[attr][fpm]['linee']
+#                 m_line.Width.values = mod_intbgd[attr][fpm]['width']
+#                 m_line.LineE.frozen = True
+#                 m_line.Width.frozen = True
+#                 m_line.norm.values = np.sum(
+#                     bgddetimsum[spec_arrinx] * np.array(
+#                         mod_intbgd[attr][fpm]['ifactors'])
+#                 ) / np.sum(bgddetimsum[spec_arrinx])
+                
```
My understanding is that this is how the IDL program behaves, which lets these three solar related lines to be free (from nuskybgd_fitab.pro, ref below).

https://github.com/NuSTAR/nuskybgd/blob/9312caf14215eb1f29987f5fe1f9722b7b2d63d8/pro/nuskybgd_fitab.pro#L278

achronal commented on nuskybgd/model.py:
```diff
@@ -523,71 +523,138 @@ def applymodel_intbgd(presets, refspec, bgddetimsum, model_num, src_number=None,
+                    this_par = {parnum:f"{mod_intbgd[attr][fpm]['linee']}, -0.1",
+                            parnum+1:f"{mod_intbgd[attr][fpm]['width']}, -0.1",
+                            parnum+2:f"{line_norm}, 0.1"}
                 else:
-                    m_line.norm.values = norm_preset
+                    rel_norm = line_norm / ref_norm                
+                    this_par = {parnum:f"{mod_intbgd[attr][fpm]['linee']}, -0.1",
+                                parnum+1:f"{mod_intbgd[attr][fpm]['width']}, -0.1",
+                                parnum+2:f"={model_name}:{ref_par}*{rel_norm}"}
+                par_dict.update(this_par)
+                parnum += 3
+
+                continue
+                
+                # Replaced by above
+#                 attr = 'lorentz_%d' % (attr_n + 1)
+#                 m_line = getattr(m, attr)
+#                 m_line.LineE.values = mod_intbgd[attr][fpm]['linee']
+#                 m_line.Width.values = mod_intbgd[attr][fpm]['width']
+#                 m_line.LineE.frozen = True
+#                 m_line.Width.frozen = True
+# 
+#                 norm_preset = np.sum(
+#                     bgddetimsum[spec_arrinx] * np.array(
+#                         mod_intbgd[attr][fpm]['ifactors'])
+#                 ) / np.sum(bgddetimsum[spec_arrinx])
+# 
+#                 if fix_line_ratios:
+#                     lorentz_5_norm_npar = spec_arrinx * m.nParameters + 16
+# 
+#                     m_line.norm.link = '%e * %s:p%d' % (
+#                         norm_preset / m.lorentz_5.norm.values[0],
+#                         model_name,
+#                         lorentz_5_norm_npar)
+#                 else:
+#                     m_line.norm.values = norm_preset
+
+            # Actually apply the changes here
+            m.setPars(par_dict)
+
```
(Note that this part concerns the line norms in the reference spectra)

I added this fix_line_ratios switch to toggle whether ref spectra lines from the 6th line onwards are linked to the 5th one, in response to Dan's request to change the default behavior to *not* tie them together. The ref_set switch here would undo that :(

achronal reviewed (changes requested):
Thanks! Very good point regarding setPars, definitely the better way. I have one small request. Some of the modifications change behavior (e.g. the line ratio values; the two blocks I added comments to) and they are best addressed independently. Please could you separate these changes into multiple pull requests, so that there is one that only concerns parameter assignment?

Not trying to have you do more work. I would like to merge your changes addressing #4 before making changes to apply it more broadly.

bwgref commented:
I'd rather not break all of the code. I'd rather just add in the functionality to choose to use the preset values or not (and I can default it to "use the preset values") rather than go through the effort of having multiple PRs.

I've allowed all of the "solar" norms to be free (as requested). I think that whole part of the model is being revisited anyway.

I've adjusted the PR to respect the fix_line_ratios toggle (assuming "False" means that are *not* fixed to be relative).

It doesn't look like there is a way to set fix_line_ratios in the current CLI calls when implementing backgrounds. And it defaults to "False". I tested this, and if you hard code fit_line_ratios=True when calling model.apply_intbgd it (correctly) sets the line normalizations to be tied together. 

You just need to update the interface to allow this to happen (i.e. figuring out whether this should be another parameter in the input JSON file or what).

bwgref committed:
Updates to how intbgd is created

Used setPars formalism here rather than accessing the parameters directly. For me this sped things up considerably (minutes --> seconds) for the creation of the bgdparams.xcm file.

Also forced all of the line strengths to be relative and tied together rather than independent.
Properly track A/B parameters.

Had to track the parameter offsets wrt to the reference spectrum when mixing FPMA and FPMB.

"Solar" low energy components to be all be free based on presets.

Respects use of fix_line_ratios flag to have the lines be tied. Default I guess is fix_line_ratios = False? (it probably should not be)

bwgref commented:
Ready to be reviewed.

achronal commented:
Thanks. I have a few comments about the points below:

> It doesn't look like there is a way to set fix_line_ratios in the current CLI calls when implementing backgrounds. And it defaults to "False". I tested this, and if you hard code fit_line_ratios=True when calling model.apply_intbgd it (correctly) sets the line normalizations to be tied together.

* The fix_line_ratios toggle here replicates the IDL program in that the IDL program had the same flag, and it is not applied by default--you'd have to set it to evaluate true when calling nuskybgd_fitab if you wanted to fix the line ratios.

> You just need to update the interface to allow this to happen (i.e. figuring out whether this should be another parameter in the input JSON file or what).

* Currently this is only available when calling the Python function. I will add this setting to the CLI interface.

* Regarding the ref_set flag you introduced--it now has the effect of copying fix_line_ratios and isn't doing anything by itself. Also, that entire block is either free or fixed (the reference component is the last one in the previous block, not the first one in that block). I will just amend this when I merge your edits.

* I will leave out the changes to the intn preset normalization values---Dan is puzzled by the 10^4 difference and suggests I find out why. Let's save that for another issue dealing with the model values.

bwgref commented:
> * Currently this is only available when calling the Python function. I will add this setting to the CLI interface.

Yes, this was my point. I would add a parser to the input json file to set this parameter there. If the command-line API doesn't allow access to undocumented flags then the flags don't exist for the end user.

> * Regarding the ref_set flag you introduced--it now has the effect of copying fix_line_ratios and isn't doing anything by itself. Also, that entire block is either free or fixed (the reference component is the last one in the previous block, not the first one in that block). I will just amend this when I merge your edits.

I'm not sure what you mean by "the reference component is the last one in the previous block". You're right, the way I've written it then the reference line will be the *first* line in each block. 

But we're now four layers deep in comments and I have no idea what you're actually asking me to do, so I'll keep this branch on my local fork for now to make it usable for me and you can revert back to your main branch and make the changes yourself.

> * I will leave out the changes to the intn preset normalization values---Dan is puzzled by the 10^4 difference and suggests I find out why. Let's save that for another issue dealing with the model values.

Okay. I'm guessing that in the python version it's doing something odd when integrating over area and/or there's an issue with how this version is handling backscal (but somehow only for this parameter). But it has to be something specific to how you're computing the normalizations for intn.

achronal commented:
> I'm not sure what you mean by "the reference component is the last one in the previous block". You're right, the way I've written it then the reference line will be the _first_ line in each block.

That will set lorentz_6 as the reference for all the lines that follow. However, lorentz_5 is the one to use (19 keV), that's in the last item in the previous for-block ('lorentz', 'lorentz_3', 'lorentz_4', 'lorentz_5').

You don't need to do anything further about this PR, I'll merge it in soon. Thanks for working on this. I just wanted to let you know about the above, in case you are currently using it and it incorrectly links the lines in the resulting model.

achronal commented:
Your changes related to #4 has been incorporated into master (https://github.com/achronal/nuskybgd-py/compare/705a1e0a2a30fcefd366f530b527ca6a58076900...master). #2 and #3 to be addressed separately.
